File: app/src/utils/migrationFailure.ts
import { invoke } from '@tauri-apps/api/core';

interface MigrationFailureMarkerFile {
  marker_path: string;
  contents: string;
}

export interface MigrationFailureFact {
  label: string;
  value: string;
}

export interface MigrationFailure {
  markerPath: string;
  facts: MigrationFailureFact[];
}

const KNOWN_FACTS: Array<[key: string, label: string]> = [
  ['daemon_log_path', 'Daemon log'],
  ['data_dir', 'Data directory'],
  ['database_path', 'Database'],
  ['backup_path', 'Backup'],
  ['schema_version_from', 'Schema version before'],
  ['schema_version_to', 'Schema version after'],
  ['error', 'Error'],
  ['binary_version', 'attn version'],
  ['failed_at', 'Failed at'],
];

function factValue(value: unknown): string {
  return typeof value === 'string' ? value : JSON.stringify(value);
}

export function migrationFailureFromMarker(file: MigrationFailureMarkerFile): MigrationFailure {
  let parsed: unknown;
  try {
    parsed = JSON.parse(file.contents);
  } catch {
    parsed = null;
  }
  if (!parsed || typeof parsed !== 'object' || Array.isArray(parsed)) {
    return {
      markerPath: file.marker_path,
      facts: [{ label: 'Marker contents (not a JSON object)', value: file.contents }],
    };
  }
  const record = parsed as Record<string, unknown>;
  const known = new Set(KNOWN_FACTS.map(([key]) => key));
  const facts = [
    ...KNOWN_FACTS
      .filter(([key]) => key in record)
      .map(([key, label]) => ({ label, value: factValue(record[key]) })),
    ...Object.keys(record)
      .filter((key) => !known.has(key))
      .map((key) => ({ label: key, value: factValue(record[key]) })),
  ];
  return { markerPath: file.marker_path, facts };
}

export async function readMigrationFailureMarker(): Promise<MigrationFailure | null> {
  try {
    const file = await invoke<MigrationFailureMarkerFile | null>('read_migration_failure');
    return file ? migrationFailureFromMarker(file) : null;
  } catch (err) {
    console.error('[Daemon] Failed to read the migration failure marker:', err);
    return null;
  }
}

File: internal/daemon/migration_failure.go
package daemon

import (
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"time"

	"github.com/victorarias/attn/internal/buildinfo"
	"github.com/victorarias/attn/internal/config"
	"github.com/victorarias/attn/internal/store"
)

const MigrationFailureFileName = "migration-failure.json"

type MigrationFailure struct {
	DaemonLogPath     string `json:"daemon_log_path"`
	DataDir           string `json:"data_dir"`
	DatabasePath      string `json:"database_path"`
	BackupPath        string `json:"backup_path"`
	SchemaVersionFrom int    `json:"schema_version_from"`
	SchemaVersionTo   int    `json:"schema_version_to"`
	Error             string `json:"error"`
	BinaryVersion     string `json:"binary_version"`
	BinaryCommit      string `json:"binary_commit"`
	FailedAt          string `json:"failed_at"`
}

func MigrationFailurePath(dataRoot string) string {
	return filepath.Join(dataRoot, MigrationFailureFileName)
}

func (d *Daemon) openStore() error {
	if d.store != nil {
		return nil
	}
	dbPath := config.DBPath()
	opened, upgrade, err := store.Open(dbPath)
	markerPath := MigrationFailurePath(d.dataRoot)
	if err != nil {
		failure := MigrationFailure{
			DaemonLogPath:     config.LogPath(),
			DataDir:           d.dataRoot,
			DatabasePath:      dbPath,
			BackupPath:        upgrade.BackupPath,
			SchemaVersionFrom: upgrade.From,
			SchemaVersionTo:   upgrade.To,
			Error:             err.Error(),
			BinaryVersion:     buildinfo.Version,
			BinaryCommit:      buildinfo.GitCommit,
			FailedAt:          time.Now().UTC().Format(time.RFC3339),
		}
		d.logf("database upgrade failed: %v", err)
		if writeErr := writeMigrationFailure(markerPath, failure); writeErr != nil {
			d.logf("could not write the migration failure marker %s: %v", markerPath, writeErr)
			return fmt.Errorf("opening the database at %s: %w (the failure marker could not be written: %v)", dbPath, err, writeErr)
		}
		return fmt.Errorf("opening the database at %s: %w (details in %s)", dbPath, err, markerPath)
	}
	d.store = opened
	if upgrade.From != upgrade.To {
		d.logf("database %s upgraded from schema v%d to v%d; backup %s", dbPath, upgrade.From, upgrade.To, upgrade.BackupPath)
	}
	if err := os.Remove(markerPath); err != nil && !errors.Is(err, os.ErrNotExist) {
		d.logf("could not remove the stale migration failure marker %s: %v", markerPath, err)
	}
	return nil
}

func writeMigrationFailure(path string, failure MigrationFailure) error {
	data, err := json.MarshalIndent(failure, "", "  ")
	if err != nil {
		return err
	}
	if err := os.MkdirAll(filepath.Dir(path), 0700); err != nil {
		return err
	}
	tmp := fmt.Sprintf("%s.tmp.%d", path, os.Getpid())
	if err := os.WriteFile(tmp, append(data, '\n'), 0600); err != nil {
		return err
	}
	if err := os.Rename(tmp, path); err != nil {
		_ = os.Remove(tmp)
		return err
	}
	return nil
}
